# CONJUNTO DE EJERCICIOS 1.3


## 1. Aritmética de corte de tres dígitos

Utilice aritmética de corte de tres dígitos para calcular las siguientes sumas. Para cada parte, ¿qué método es más preciso y por qué?

### a. 


$$
\sum_{i=1}^{10} \frac{1}{i^2} 
\
     Primero por:   
\[
1 + \frac{1}{4} + \ldots + \frac{1}{100} 
\]
    y luego por:
\[
\frac{1}{100} + \frac{1}{81} + \ldots + 1
\$$




In [5]:
def truncar(valor: float, decimales: int) -> float:
    factor = 10 ** decimales
    return int(valor * factor) / factor

def suma_Uno_mas(n: int) -> float:
    primer = 0.0  # Inicializar la suma en 0
    for i in range(1, n + 1):  # Iterar desde 1 hasta n
        term = truncar(1 / (i ** 2), 3)  # Truncar 1/i^2 a 3 cifras decimales
        primer += term  # Sumar el término truncado
    return primer  # Devolver la suma

a = suma_Uno_mas(10)  
print(a)  # Mostrar el resultado


1.547


In [6]:
def suma_al_reves(n: int) -> float:
    suma = 0.0  # Inicializar la suma en 0
    for i in range(n, 0, -1):  # Desde n hasta 1
        term = truncar(1 / (i ** 2), 3)  # Truncar 1/i^2 a 3 cifras decimales
        suma += term  # Sumar el término truncado
    return suma  # Devolver la suma

resultado = suma_al_reves(10)
print(resultado)  # Mostrar el resultado


1.547


### b. 
$$\
\sum_{i=1}^{10} \frac{1}{i^3}
\]
Primero por:
\[
1 + \frac{1}{8} + \frac{1}{27} + \ldots + \frac{1}{1000}
\]
y luego por:
\[
\frac{1}{1000} + \frac{1}{729} + \ldots + 1
\$$


In [10]:


def suma_Cubica(n: int) -> float:
    primer = 0.0  # Inicializar la suma en 0
    for i in range(1, n + 1):  # Iterar desde 1 hasta n
        term = truncar(1 / (i ** 3), 3)  # Truncar 1/i^2 a 3 cifras decimales
        primer += term  # Sumar el término truncado
    return primer  # Devolver la suma

a = suma_Cubica(10)  
print(a)
def suma_al_reves(n: int) -> float:
    suma = 0.0  # Inicializar la suma en 0
    for i in range(n, 0, -1):  # Desde n hasta 1
        term = truncar(1 / (i ** 3), 3)  # Truncar 1/i^3 a 3 cifras decimales
        suma += term  # Sumar el término truncado
    return suma  # Devolver la suma

resultado = suma_al_reves(10)
print(resultado)  # Mostrar el resultado



1.1939999999999995
1.194


El método más preciso es el de la suma directa. Esto se debe a que:

La suma directa mantiene la precisión al sumar términos más grandes primero, lo que minimiza la pérdida de significancia en las cifras cuando se suman términos más pequeños.
En la suma inversa, los errores de redondeo son más propensos a afectar el resultado final debido a la naturaleza de los números involucrados.

## 2. Serie de Maclaurin para la función arcotangente

La serie de Maclaurin para la función arcotangente converge para $$\-1 < x \leq 1\$$ y está dada por:

$$
\arctan x = \lim_{n \to \infty} P_n(x) = \lim_{n \to \infty} \sum_{i=1}^{n} (-1)^{i+1} \frac{x^{2i-1}}{2i-1}
$$

### a.
Utilice el hecho de que $$\tan \frac{\pi}{4} = 1\ $$
para determinar el número \(n\) de términos de la serie que se necesita sumar para garantizar que $$\|4P_n(1) - \pi| < 10^{-3}\$$.




## Pasos para resolver el problema:

1. **Definir la serie de Maclaurin**

    para $$ arctan(1) $$ :

   Para \( x = 1 \), la serie se convierte en:

   $$ P_n(1) = \sum_{i=1}^n (-1)^{i+1} \frac{1^{2i-1}}{2i-1} = \sum_{i=1}^n (-1)^{i+1} \frac{1}{2i-1}
    $$

   Esto se debe a que $$\ 1^{2i-1} = 1 \$$ para cualquier i .

3. **Condición de precisión**:

   La serie $$ P_n(1) $$ se multiplica por 4 para aproximar \(\pi\). Entonces,  encontrar un  n tal que:

  $$
   \left| 4 P_n(1) - \pi \right| < 10^{-3}
$$

5. **Cálculo de \( n \)**:

   Utilizaremos un bucle para calcular la suma hasta que la diferencia con π sea menor a $$\ 10^{-3} $$.




In [1]:

import math

def calcular_n(precision: float) -> int:
    pi_approx = 0.0
    n = 0
    x = 1  # Definiendo el valor de x según el problema
    while True:
        n += 1
        term = ((-1)**(n + 1)) * (x**(2 * n - 1) / (2 * n - 1))  # Término de la serie
        pi_approx += term
        if abs(4 * pi_approx - math.pi) < precision:
            break
    return n

n_necesario = calcular_n(10**-3)
print(n_necesario)






1000


### b.
El lenguaje de programación C++ requiere que el valor de π se encuentre dentro de \(10^{-10}\). ¿Cuántos términos de la serie se necesitarían sumar para obtener este grado de precisión?


In [ ]:
nCplusplus=calcular_n(10**-10)
print(nCplusplus)

## 3. Otra fórmula para calcular \(\pi\)

Otra fórmula para calcular \(\pi\) se puede deducir a partir de la identidad:
$$
\frac{\pi}{4} = 4 \text{arctan } \frac{1}{5} - \text{arctan } \frac{1}{239}
\$$
Determine el número de términos que se deben sumar para garantizar una aproximación de π dentro de \(10^{-3}\).



### Resolución del ejercicio 3

#### Identidad dada:
Se nos proporciona la siguiente identidad para calcular π:

$$
\frac{\pi}{4} = 4 \arctan\left(\frac{1}{5}\right) - \arctan\left(\frac{1}{239}\right)
\$$

Esta fórmula se conoce como una fórmula de Machin para aproximar el valor de π. La idea es usar la serie de Maclaurin para \(\arctan(x)\) y determinar cuántos términos debemos sumar para alcanzar una precisión de \(10^{-3}\).

#### Serie de Maclaurin para \(\arctan(x)\):
La serie de Maclaurin para la función \(\arctan(x)\) está dada por:

$$
\arctan(x) = \sum_{i=0}^n (-1)^i \frac{x^{2i+1}}{2i+1}
\$$

#### Expresión de π:
Utilizando la identidad dada, podemos escribir la expresión para π como:

$$
\pi \approx 4 \left[ 4 \sum_{i=0}^n (-1)^i \frac{\left(\frac{1}{5}\right)^{2i+1}}{2i+1} - \sum_{i=0}^n (-1)^i \frac{\left(\frac{1}{239}\right)^{2i+1}}{2i+1} \right]
\$$

Nuestro objetivo es determinar cuántos términos (\(n\)) debemos sumar para que el valor calculado de \(\pi\) esté dentro de una precisión de \(10^{-3}\), es decir, cumplir con la siguiente condición:

$$
\left|\pi_{\text{aprox}} - \pi\right| < 10^{-3}
\$$

In [ ]:
import math

def calcular_n(precision: float) -> int:
    pi_approx = 0.0
    n = 0
    while True:
        n += 1
        term1 = ((-1) ** (n + 1)) * (1 / 5) ** (2 * n - 1) / (2 * n - 1)
        term2 = ((-1) ** (n + 1)) * (1 / 239) ** (2 * n - 1) / (2 * n - 1)
        
        # Aproximación a pi usando la fórmula de Machin
        pi_approx = 4 * (4 * term1 - term2)
        
        # Comprobar si la aproximación de pi cumple con la precisión requerida
        if abs(pi_approx - math.pi) < precision:
            break
    return n

# Determinar el número de términos necesarios para una precisión de 10^(-3)
n_necesario = calcular_n(10**-3)
print(f"Número de términos necesarios: {n_necesario}")




## 5. Cálculos de suma

### a.
¿Cuántas multiplicaciones y sumas se requieren para determinar una suma de la forma 
$$\
\sum_{i=1}^{n} \sum_{j=1}^{n} a_i b_j?
\$$




### Resolución del ejercicio 5

Queremos determinar cuántas multiplicaciones y sumas se requieren para calcular la siguiente expresión:

\[
\sum_{i=1}^n \sum_{j=1}^i a_i b_j
\]

#### Análisis:

- **Bucle externo**: Recorre el índice \( i \) desde \( 1 \) hasta \( n \).
- **Bucle interno**: Recorre el índice \( j \) desde \( 1 \) hasta \( i \).

En cada iteración del bucle interno, se realiza una multiplicación \( a_i \cdot b_j \) y luego se suma el resultado al total acumulado.

#### Total de operaciones:

1. **Multiplicaciones**: Para cada \( i \), se realizan \( i \) multiplicaciones, por lo que el total de multiplicaciones es:

   \[
   1 + 2 + 3 + \ldots + n = \frac{n(n+1)}{2}
   \]

2. **Sumas**: En cada iteración del bucle interno se suma el resultado de la multiplicación al total acumulado, lo que implica \( \frac{n(n+1)}{2} \) sumas.

#### Conclusión:

Para determinar una suma de la forma \(\sum_{i=1}^n \sum_{j=1}^i a_i b_j\), se requieren:

- \(\frac{n(n+1)}{2}\) multiplicaciones.
- \(\frac{n(n+1)}{2}\) sumas.


In [ ]:
def contar_operaciones(n):
    # Cálculo del número de multiplicaciones y sumas
    num_multiplicaciones = (n * (n + 1)) // 2
    num_sumas = num_multiplicaciones  # La cantidad de sumas es igual a la cantidad de multiplicaciones

    return num_multiplicaciones, num_sumas

# Ejemplo de uso
n = 5
multiplicaciones, sumas = contar_operaciones(n)
print(f"Para n = {n}:")
print(f"Multiplicaciones necesarias: {multiplicaciones}")
print(f"Sumas necesarias: {sumas}")


### b.
Modifique la suma en la parte a) a un formato equivalente que reduzca el número de cálculos.


### Resolución del ejercicio 5

Queremos determinar cuántas multiplicaciones y sumas se requieren para calcular la siguiente expresión:

\[
\sum_{i=1}^n \sum_{j=1}^i a_i b_j
\]

#### Análisis:

- **Bucle externo**: Recorre el índice \( i \) desde \( 1 \) hasta \( n \).
- **Bucle interno**: Recorre el índice \( j \) desde \( 1 \) hasta \( i \).

En cada iteración del bucle interno, se realiza una multiplicación \( a_i \cdot b_j \) y luego se suma el resultado al total acumulado.

#### Total de operaciones:

1. **Multiplicaciones**: Para cada \( i \), se realizan \( i \) multiplicaciones, por lo que el total de multiplicaciones es:

   \[
   1 + 2 + 3 + \ldots + n = \frac{n(n+1)}{2}
   \]

2. **Sumas**: En cada iteración del bucle interno se suma el resultado de la multiplicación al total acumulado, lo que implica \( \frac{n(n+1)}{2} \) sumas.

#### Conclusión:

Para determinar una suma de la forma \(\sum_{i=1}^n \sum_{j=1}^i a_i b_j\), se requieren:

- \(\frac{n(n+1)}{2}\) multiplicaciones.
- \(\frac{n(n+1)}{2}\) sumas.



In [3]:
#### Código en Python:


def contar_operaciones(n):
    # Cálculo del número de multiplicaciones y sumas
    num_multiplicaciones = (n * (n + 1)) // 2
    num_sumas = num_multiplicaciones  # La cantidad de sumas es igual a la cantidad de multiplicaciones

    return num_multiplicaciones, num_sumas

# Ejemplo de uso
n = 5
multiplicaciones, sumas = contar_operaciones(n)
print(f"Para n = {n}:")
print(f"Multiplicaciones necesarias: {multiplicaciones}")
print(f"Sumas necesarias: {sumas}")


Para n = 5:
Multiplicaciones necesarias: 15
Sumas necesarias: 15


## DISCUSIONES


### 2.
Las ecuaciones (1.2) y (1.3) en la sección 1.2 proporcionan formas alternativas para las raíces \(x_1\) y \(x_2\) de 
\[
ax^2 + bx + c = 0.
\]
Construya un algoritmo con entrada \(a, b, c\) y salida \(x_1, x_2\) que calcule las raíces \(x_1\) y \(x_2\) (que pueden ser iguales con conjugados complejos) mediante la mejor fórmula para cada raíz.


